### PR TITLE
Add preference lookups by height to the consensus interface

### PR DESCRIPTION
**Pull request by StephenButtolph**
## Why this should be merged

The bubbling votes optimization requires the ability to lookup the preferred value at a given height.

## How this works

In the same locations that we currently populate the `preferredIDs` set - we now populate a new map `preferredHeights` as well.

## How this was tested

Added checks to various existing preference unit tests.

### CODE REVIEW

File: snow/consensus/snowman/topological.go [hunk 1]
```diff
@@ -288,23 +308,28 @@ func (ts *Topological) RecordPoll(ctx context.Context, voteBag bag.Bag[ids.ID])
...
 	}
 	// Traverse from the preferred ID to the preferred child until there are no
 	// children.
-	for block := startBlock; block.sb != nil; block = ts.blocks[ts.tail] {
+	for block := startBlock; block.sb != nil; {
 		ts.tail = block.sb.Preference()
 		ts.preferredIDs.Add(ts.tail)
+		block = ts.blocks[ts.tail]
+		ts.preferredHeights[block.blk.Height()] = ts.tail
 	}
```
Comment (StephenButtolph) [inline snow/consensus/snowman/topological.go:333]: This is really the only subtle change - `block.sb` is only initialized after a child block is inserted. So, we are guaranteed that `block.blk` is non-nil after updating to `ts.blocks[ts.tail]`.

File: snow/consensus/snowman/topological.go [hunk 2]
```diff
@@ -251,7 +271,7 @@ func (ts *Topological) Preference() ids.ID {
 // unsuccessful poll on that block and every descendant block.
 //
 // The complexity of this function is:
-// - Runtime = 3 * |live set| + |votes|
+// - Runtime = 4 * |live set| + |votes|
```
Comment (StephenButtolph) [inline snow/consensus/snowman/topological.go:274]: Because of the additional map iteration caused by `maps.Clear`

**Review by gyuho — APPROVED** [review]

File: snow/consensus/snowman/topological.go (hunk 1 above)
Comment (StephenButtolph) [reply snow/consensus/snowman/topological.go]: Added as a comment

**Review by aaronbuchwald — APPROVED** [review]
LGTM

File: snow/consensus/snowman/topological.go [hunk 3]
```diff
@@ -301,23 +321,31 @@ func (ts *Topological) RecordPoll(ctx context.Context, voteBag bag.Bag[ids.ID])
 		return nil
 	}
 
-	// Runtime = |live set| ; Space = Constant
+	// Runtime = 2 * |live set| ; Space = Constant
```
Comment (joshua-kim) [inline snow/consensus/snowman/topological.go:324]: irrelevant comment but sometimes i feel like the comments in the code are overly verbose haha...

**Review by joshua-kim — APPROVED** [review]

File: snow/consensus/snowman/topological.go [hunk 4]
```diff
@@ -68,6 +68,10 @@ type Topological struct {
 	// preferredIDs stores the set of IDs that are currently preferred.
 	preferredIDs set.Set[ids.ID]
 
+	// preferredHeights maps heights to the set of IDs that are currently
+	// preferred.
+	preferredHeights map[uint64]ids.ID // height -> blockID
```
Comment (danlaine) [inline snow/consensus/snowman/topological.go:73]:
```suggestion
	// preferredHeights maps a height to the currently preferred 
	// block ID at that height.
	preferredHeights map[uint64]ids.ID // height -> blockID
```

**Review by danlaine — APPROVED** [review]

File: snow/consensus/snowman/topological.go (hunk 3 above)
Comment (danlaine) [reply snow/consensus/snowman/topological.go:324]: idk I feel like for a hot path where we care about the performance a lot it's nice

**Pull request merged**